### PR TITLE
[Unstructured prune]Support block sparse training

### PR DESCRIPTION
**Pull request by minghaoBD**
We are supporting block sparse in this PR.

1. By block sparse, we mean treat each MxN block in weight matrix as a group when pruning. The group ether has all zeros inside or original values. Specifically, unstructured pruning is block sparse when M=N=1. In practice, you could easily pass in a List<Integer> with 2 values to control M and N

2. Furthermore, this PR adds the **bias** as the default skipped parameters, which only accounts for small amount in the network, but may contribute a lot to the final accuracy.

3. There is one issue left in this PR: the training becomes fairly slow without GMP, when (m!=1 or n!=1). This is due to the nested for loop in _cal_mxn_avg_matrix. We might use **img2col+avg+col2img** to give it an optimization. But some corner cases and op types should be paid careful attention to.

### CODE REVIEW

File: demo/dygraph/unstructured_pruning/train.py [hunk 1]
```diff
@@ -35,6 +35,7 @@
 add_arg('threshold',            float,  0.01,               "The threshold to set zeros. Default: 0.01")
 add_arg('num_epochs',       int,  120,               "The number of total epochs. Default: 120")
 parser.add_argument('--step_epochs', nargs='+', type=int, default=[30, 60, 90], help="piecewise decay step")
+parser.add_argument('--sparse_block', nargs='+', type=int, default=[1, 1], help="The values inside the block are either all zeros or the original. [1, 1] means unstructured pruning")
```
Comment (wanghaoshuang) [inline demo/dygraph/unstructured_pruning/train.py]: 没解释清楚“sparse_block”什么意思；
没有中文文档里解释的好。

File: docs/zh_cn/api_cn/dygraph/pruners/unstructured_pruner.rst [hunk 2]
```diff
@@ -40,11 +40,14 @@ UnstructuredPruner
...
           if type(sub_layer).__name__.split('.')[-1] in NORMS_ALL:
               skip_params.add(sub_layer.full_name())
+          for param in sub_layer.parameters(include_sublayers=False):
+              if len(param.shape) == 1: skip_params.add(param.name)
       return skip_params
 
 ..
 
 - **local_sparsity(bool)** - 剪裁比例（ratio）应用的范围：local_sparsity 开启时意味着每个参与剪裁的参数矩阵稀疏度均为 'ratio'， 关闭时表示只保证模型整体稀疏度达到'ratio'，但是每个参数矩阵的稀疏度可能存在差异。
+- **sparse_block(Array<Integer>)** - 一个含有两个正整数的数组，定义了稀疏化时候block的大小。即 sparse_block[0] x sparse_block[1]中的参数作为一个整体，要么被稀疏置0，要么保持不变。默认为 [1,1]，代表非结构化稀疏。
```
Comment (wanghaoshuang) [inline docs/zh_cn/api_cn/dygraph/pruners/unstructured_pruner.rst]: >即 sparse_block[0] x sparse_block[1]中
"sparse_block[0] x sparse_block[1]"是指block的shape，不能说“在shape中...”

>要么被稀疏置0，要么保持不变

“被稀疏置0”改成“被置0”？

File: paddleslim/dygraph/prune/unstructured_pruner.py [hunk 3]
```diff
@@ -97,6 +120,15 @@ def update_threshold(self):
                     continue
                 t_param = param.value().get_tensor()
                 v_param = np.array(t_param)
+
+                if (self.sparse_block[0] * self.sparse_block[1] / v_param.size
+                        >= 0.05):
```
Comment (wanghaoshuang) [inline paddleslim/dygraph/prune/unstructured_pruner.py]: 在当前文件头几行，声明一个意义明显的全局变量来表示0.05这个数值吧。

File: paddleslim/dygraph/prune/unstructured_pruner.py [hunk 4]
```diff
@@ -97,6 +120,15 @@ def update_threshold(self):
...
+                    print(
+                        "Your sparse block size {} might be too large for the param {} with shape {}, the sparsity of this param might not be precise. Please decrease your sparse block size if possible.".
```
Comment (wanghaoshuang) [inline paddleslim/dygraph/prune/unstructured_pruner.py]: 明确告诉用户阈值是多少？要不用户要尝试多次才可以。

File: paddleslim/prune/unstructured_pruner.py [hunk 5]
```diff
@@ -108,6 +119,17 @@ def _apply_masks(self, program, mask_func):
             d_masks[_param.name] = _mask.name
         return d_masks
 
+    def _cal_mxn_avg_matrix(self, mat, m=1, n=1):
```
Comment (wanghaoshuang) [inline paddleslim/prune/unstructured_pruner.py]: 把这个方法抽出来作为公共方法？

File: paddleslim/prune/unstructured_pruner.py [hunk 6]
```diff
@@ -161,6 +183,13 @@ def update_threshold(self):
                 continue
             t_param = self.scope.find_var(param).get_tensor()
             v_param = np.array(t_param)
+            if (self.sparse_block[0] * self.sparse_block[1] / v_param.size >=
+                    0.05):
```
Comment (wanghaoshuang) [inline paddleslim/prune/unstructured_pruner.py]: 是否能确保动静态图中这个阈值保持一致？

File: paddleslim/prune/unstructured_pruner.py [hunk 7]
```diff
@@ -161,6 +183,13 @@ def update_threshold(self):
...
+                print(
```
Comment (wanghaoshuang) [inline paddleslim/prune/unstructured_pruner.py:178]: 明确告知用户阈值是多少，或者用assert语句，用户也能看到具体报错的判定条件。

File: demo/dygraph/unstructured_pruning/train.py (hunk 1 above)
Comment (minghaoBD) [reply demo/dygraph/unstructured_pruning/train.py]: There must be two integers inside this array. The array defines the shape of the block, the values within which are either sparsified to all zeros or kept original. [1, 1] means unstructured pruning. Default: [1, 1]

File: docs/zh_cn/api_cn/dygraph/pruners/unstructured_pruner.rst (hunk 2 above)
Comment (minghaoBD) [reply docs/zh_cn/api_cn/dygraph/pruners/unstructured_pruner.rst]: DONE

File: paddleslim/dygraph/prune/unstructured_pruner.py (hunk 3 above)
Comment (minghaoBD) [reply paddleslim/dygraph/prune/unstructured_pruner.py]: BLOCK_SPARSE_ACCURATE_THRESHOLD=0.05

File: paddleslim/dygraph/prune/unstructured_pruner.py (hunk 4 above)
Comment (minghaoBD) [reply paddleslim/dygraph/prune/unstructured_pruner.py]: image

File: paddleslim/prune/unstructured_pruner.py (hunk 5 above)
Comment (minghaoBD) [reply paddleslim/prune/unstructured_pruner.py]: 已经抽出来，和BLOCK_SPARSE_ACCURATE_THRESHOLD一起放在paddleslim.prune.unstructured_prune_utils文件下。动静态图共同调用。

File: paddleslim/prune/unstructured_pruner.py (hunk 7 above)
Comment (minghaoBD) [reply paddleslim/prune/unstructured_pruner.py:178]: 同上

File: paddleslim/prune/unstructured_pruner.py (hunk 6 above)
Comment (minghaoBD) [reply paddleslim/prune/unstructured_pruner.py]: 同上

**Review by wanghaoshuang — DISMISSED** [review]

**Review by wanghaoshuang — APPROVED** [review]
